### ISSUE TITLE
PRQL dialect as external contribution?

### ISSUE THREAD

**Issue by max-sixty** [ClickHouse/ClickHouse#46182]
**Use case**

Hi from the team at [PRQL](github.com/prql/prql)!

I was pointed towards https://github.com/ClickHouse/ClickHouse/issues/42194, which has this intern task:

> Support for PRQL as a query language in ClickHouse
>
> If you like programming languages.
>
> After the brilliant work of Denis Bolonin, ClickHouse can statically link Rust libraries: https://clickhouse.com/docs/en/development/integrating_rust_libraries/
>
> PRQL https://prql-lang.org/ is a declarative language similar to SQL, implemented as a Rust library. We can plug it into ClickHouse as a dialect.

**Describe the solution you'd like**

It looks like no intern grabbed the project (and maybe the internships are over by now?).

_Would you be up for a contribution from an external contributor?_ If so, we'll add an issue at PRQL and hopefully it will get picked up there.

**Additional context**

For context — we're trying to be as open and welcoming as possible for projects to integrate with. In particular, from https://prql-lang.org:

> PRQL will always be fully open-source and will never have a commercial product.

**Comment by seshWCS** [comment]
I will compile prql-query to sql, then I'll send it to executeQuery function

**Comment by seshWCS** [comment]
Now I integrated prql-lib to repo and added ffi to it.

**Comment by alexey-milovidov** [comment]
@max-sixty 

@antaljanosbenjamin will help to finish the contribution.
He has a few questions, for example:

> there is some issue with the prql module where it import some std lib and it [triggers the memory sanitizer](https://s3.amazonaws.com/clickhouse-test-reports/50050/db86d940989e482fd7b7f79e68aa6cb5603ab991/fuzzer_astfuzzermsan/report.html)

Does PRQL use native (unsafe) code?

**Comment by max-sixty** [comment]
Hi @alexey-milovidov ! That's great news. Thanks @antaljanosbenjamin .



> there is some issue with the prql module where it import some std lib and it [triggers the memory sanitizer](https://s3.amazonaws.com/clickhouse-test-reports/50050/db86d940989e482fd7b7f79e68aa6cb5603ab991/fuzzer_astfuzzermsan/report.html)

How is the code on the ClickHouse side implemented currently? We have no unsafe code in our main rust crate, `prql-compiler`. We have some unsafe code in our C bindings, `prql-lib`, which I think is unavoidable given the C bindings.

I'm guessing you're quoting from something internal, since I don't see that discussion. Happy to work through any issues if you share them here.

**Comment by max-sixty** [comment]
I guess that was started a while ago. It's using quite an old version of `prql-lib`, which IIUC has been vendored in.

It's might be worth updating to the latest version — the C code in `prql-lib` is much better quality, the dependencies are more up to date, and it's also a recent version of PRQL, so there's a decent chance it solves one or both of the memory issues.

We enforce better API stability, so it won't require more intrusive changes in the future. Here's the README: https://github.com/PRQL/prql/blob/8ecfdebedb031042b78d9ecfbb83910a5db1494a/bindings/prql-lib/README.md.

Would that be a tractable change from your perspective?

I'm also CCing @aljazerzen, who is on the core PRQL team and wrote most of `prql-lib`, in case he has any thoughts.

**Comment by alexey-milovidov** [comment]
Good. Just in case we build every dependency from the source, the versions are pinned (updated on need), and there is no requirement for API stability; there is also no use for installed/packaged libraries. From our standpoint, using the Rust library directly maybe even more preferred than using the C library.

**Comment by max-sixty** [comment]
If you can use the `prql-compiler` crate directly, rather than `prql-lib`, then that's even easier. Can you take a rust dependency directly?

Currently it looks like `prql-lib` is vendored in. For context, `prql-lib`'s only role is to offer C bindings for `prql-compiler`.

**Comment by antaljanosbenjamin** [comment]
Yes, currently we use the [handwritten](https://github.com/ClickHouse/ClickHouse/blob/prql-integration/rust/prql/src/lib.rs) wrapper around `prql-compiler` that directly calls `prql_compiler::compile`. The reason for this I think is we already have integration for Rust libraries, but not with go libraries.

**Comment by max-sixty** [comment]
Ok, PRQL's compiler is a rust crate: https://crates.io/crates/prql-compiler. (It is not a `go` library).

If ClickHouse supports rust integrations, could this be as simple as `use prql_compiler; prql_compiler::compile(...)`?

Possibly the confusion was that `prql-lib` is an approach for `go` _consume_ `prql-compiler`...

Does that make sense? Thanks for your efforts so far and let me know what I can do to help. Sorry if there was some rust / `go` confusion prior.

**Comment by antaljanosbenjamin** [comment]
Everything okay for now. As Alexey mentioned we build everything from source and as far as I see we don't have go libraries that we build as part of ClickHouse. Therefore we opted not to use `prql-lib`, but to write a simple C wrapper around the `prql-compiler` crate (more precisely around `prql_compiler::compile`.

Right now I am dealing with some issues regarding out CI, but hopefully that is resolved and we just have to wait for the build and checks to finish. I updated the version of `prql-compiler` to `0.8.1`, hopefully that will solve the memory sanitizer issue. If not, then I will deep dive into what is happening under the hood.

Either way, I will let you konw if I need any help.
